### PR TITLE
feat: Add `signature_verification_result` to schnorr stdlib

### PR DESCRIPTION
**Pull request by phated**
# Description

Noir needs `verify_signature` in the schnorr stdlib to return a `bool_t` indicating whether the signature was valid or not.

Currently a draft because I tried to update @kevaundray's code to add a constraint but I have no idea if I did it correctly.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/aztec/stdlib/encryption/schnorr/schnorr.cpp
Comment (codygunton) [inline cpp/src/aztec/stdlib/encryption/schnorr/schnorr.cpp:296]: `assert_equal` imposes a constraint (a copy constraint, which is free in the sense of adding no gates) that two things are equal. Your change breaks the model of this function. As it was before, the circuit would be supplied with non-satisfying witnesses when the signature verification fails, i.e., the prover would produce a proof that would fail to verify. With your change, the prover would supply a passing proof that the signature failed to verify. This is why some of the tests don't pass. To resolve, you could put the `assert_equal`'s back in place.

Comment (codygunton) [reply cpp/src/aztec/stdlib/encryption/schnorr/schnorr.cpp:296]: @kevaundray Are you proposing to change the semantics here? Is there a use case for proving that a signature does _not_ verify?

**Review by codygunton — CHANGES_REQUESTED** [review]

File: cpp/src/aztec/stdlib/encryption/schnorr/schnorr.cpp
Comment (zac-williamson) [reply cpp/src/aztec/stdlib/encryption/schnorr/schnorr.cpp:296]: One thing to note is that if the current behaviour has a minor issue:

It is not currently possible to distinguish between the signature being invalid, and a Prover deliberately providing incorrect witness values.  

This isn't a problem for the Aztec Connect use case, but it may be an issue for a Noir app where proof computation is being delegated to a third party (i.e. you can't distinguish between "I can't make a valid proof because you gave me a bad signature" and "I *won't* make a valid proof because I'm dishonest"

Comment (phated) [reply cpp/src/aztec/stdlib/encryption/schnorr/schnorr.cpp:296]: I've re-added the `assert_equal` calls, which I mistakenly removed when I made the change to the logic to add actual constraints (I was confused by the "equal" name and "==" (equal operator) and thought they were the same thing).

I'm pretty sure Noir just needs to be able to know the that something was valid/invalid via return type (and if I understand correctly, that needs to be constrained)

Comment (codygunton) [reply cpp/src/aztec/stdlib/encryption/schnorr/schnorr.cpp:296]: @zac-williamson so perhaps remove the `assert_equal`s and rename to `verify_signature_unsafe` or `prove_signature_verification_result` or something like this?

**Review by zac-williamson — DISMISSED** [review]
I can see how it would be useful to have a verification method that does not assert but instead returns a boolean. It gives more flexibility to developers and makes it easier to evaluate branching paths where one wishes to conditionally verify a signature.

We have a lot of tests and circuit code that relies on the signature verification algorithm to produce unsatisfiable constraints if the signature is invalid.

I think the fastest path forward is to create a new independent method that returns a bool_t, which does not interfere with existing tests.

**Comment by codygunton** [comment]
Rebased and resolved conflicts with `waffle` removal PR.

**Review by codygunton — APPROVED** [review]

**Comment by codygunton** [comment]
@zac-williamson looks good? Merging is blocked until you approve :innocent:

**Comment by codygunton** [comment]
^ rebase

**Pull request merged**
